logstash-core/src/main/java/org/logstash/config/ir/PipelineConfig.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *	http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.logstash.config.ir;
+
+import org.apache.commons.codec.digest.DigestUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jruby.*;
+import org.jruby.runtime.builtin.IRubyObject;
+import org.logstash.common.IncompleteSourceWithMetadataException;
+import org.logstash.common.SourceWithMetadata;
+
+import java.time.LocalDateTime;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static org.logstash.RubyUtil.RUBY;
+
+public final class PipelineConfig {
+
+    private static class LineToSource {
+        private final int startLine;
+        private final int endLine;
+        private final SourceWithMetadata source;
+
+        LineToSource(int startLine, int endLine, SourceWithMetadata source) {
+            this.startLine = startLine;
+            this.endLine = endLine;
+            this.source = source;
+        }
+
+        boolean includeLine(int lineNumber) {
+            return startLine <= lineNumber && lineNumber <= endLine;
+        }
+    }
+
+    private static final Logger logger = LogManager.getLogger(PipelineConfig.class);
+
+    private RubyClass source;
+    private String pipelineId;
+    private List<SourceWithMetadata> confParts;
+    private RubyObject settings;
+    private LocalDateTime readAt;
+    private String configHash;
+    private String configString;
+    private List<LineToSource> sourceRefs;
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    public PipelineConfig(RubyClass source, RubySymbol pipelineId, RubyObject uncastedConfigParts, RubyObject logstashSettings) {
```

Comment:
Note for reviewers:
The PipelineConfig constructor is intended to accept ideally `uncastedConfigParts` as List of SourceWithMetadata, but to be transparent to JRuby it brings RubyObject because previous Ruby code accepted single object or arrays and in the first case the initialize wrapped it into an array